### ISSUE TITLE
Unable to send to graphite

### ISSUE THREAD

**Issue by grantis** [sitespeedio/sitespeed.io#2894]
Hey guys, 

I've setup my sitespeed with graphite and grafana and im trying to send the data to graphite and getting a look up error. Its quite odd. 

```
platform: linux
image_resource:
  type: docker-image
  source:
    repository: sitespeedio/sitespeed.io
    aws_access_key_id: ((aws-access-key-id))
    aws_secret_access_key: ((aws-secret-access-key))
caches:
  - path: root/.npm
inputs:
  - name: source-code
run:
  dir: source-code/test/perf
  user: root
  path: sh
  args:
  - -xce
  - |

    GRAPHITE_HOST=https://graphite-us-central1.grafana.net/metrics
    GRAPHITE_USER=username
    GRAPHITE_ACCESS_KEY=password

    npm i
    /usr/src/app/docker/scripts/start.sh --graphite.host $GRAPHITE_HOST --graphite.port 2003 --graphite.auth $GRAPHITE_USER:$GRAPHITE_ACCESS_KEY --graphite.namespace performance.test ../../test/perf/perf-test.js --multi --spa --video false --browsertime.iterations 1 --browsertime.skipHar --browsertime.cacheClearRaw

```

The tests run fine but when it trys to send to graphite i get:

```
Google Chrome 79.0.3945.79 
Mozilla Firefox 71.0
[2020-02-24 20:25:50] INFO: Running tests using Chrome - 1 iteration(s)
[2020-02-24 20:26:11] INFO: Start to measure login
[2020-02-24 20:26:12] INFO: Logging in...
[2020-02-24 20:26:14] INFO: Logged in, delay: 2.27 sec
[2020-02-24 20:27:20] WARN: BrowserError: Failed to take screenshot
    at SeleniumRunner.takeScreenshot (/usr/src/app/node_modules/browsertime/lib/core/seleniumRunner.js:259:13)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
[2020-02-24 20:27:20] INFO: Login measured
[2020-02-24 20:27:28] ERROR: Error: getaddrinfo ENOTFOUND https://graphite-us-central1.grafana.net/metrics
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:60:26)
[2020-02-24 20:27:28] ERROR: Error: getaddrinfo ENOTFOUND https://graphite-us-central1.grafana.net/metrics
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:60:26)
[2020-02-24 20:27:28] ERROR: Error: getaddrinfo ENOTFOUND https://graphite-us-central1.grafana.net/metrics
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:60:26)
[2020-02-24 20:27:28] ERROR: Error: getaddrinfo ENOTFOUND https://graphite-us-central1.grafana.net/metrics
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:60:26)
[2020-02-24 20:27:31] INFO: HTML stored in /tmp/build/e55deab7/source-code/test/perf/sitespeed-result/perf-test_js/2020-02-24-20-25-50
failed
```

I believe this could be user error and i've made a mistake somewhere in configuring. Any help or assistance would be appreciated.

**Comment by soulgalore** [comment]
Hi @grantis so you run in Docker right? I wonder it could be a DNS thing in the container? https://docs.docker.com/v17.09/engine/userguide/networking/default_network/configure-dns/

**Comment by grantis** [comment]
Hi @soulgalore , yes that's right. I'm running with both docker locally and with our ci concourse (OCI container) which offers the ability to execute tasks. Both have the same problem. I'll dig deeper on the DNS. Thanks

**Comment by grantis** [comment]
Hi @soulgalore, just wanted to clarify how I'm running sitespeed, im using the sitespeedio/sitespeed.io docker image and im passing in graphite arguments, the graphite host has been provided by grafana. 

I changed the port from 2003 to 80 and seems to connect to graphite, but im not seeing the results in grafana, maybe there is something i've missed 😕 

Here's how I'm calling it:
```
docker run --rm -it 
-v "$(pwd)":/sitespeed.io sitespeedio/sitespeed.io 
--graphite.host graphite-us-central1.grafana.net 
--graphite.auth username:secret 
--graphite.port 80 
--graphite.namespace sitespeed 
--name sitespeed 
--grafana.host https://host.grafana.net 
--grafana.auth secret  
./test/perf/perf-test.js --multi 
--spa --video false 
--browsertime.iterations 1 --browsertime.screenshot false
```
```
Google Chrome 79.0.3945.79 
Mozilla Firefox 71.0
[2020-02-26 22:34:33] INFO: Versions OS: linux 4.19.76-linuxkit nodejs: v12.13.0 sitespeed.io: 11.9.3 browsertime: 7.8.3 coach: 4.3.0
[2020-02-26 22:34:34] INFO: Running tests using Chrome - 1 iteration(s)
[2020-02-26 22:34:35] INFO: Navigating to url http://test.product.com iteration 1
[2020-02-26 22:34:51] INFO: Start to measure login
[2020-02-26 22:34:52] INFO: Logging in...
[2020-02-26 22:34:56] INFO: Logged in, delay: 4.71 sec
[2020-02-26 22:34:57] INFO: Login measured
[2020-02-26 22:35:00] INFO: https://test.product.com/ 36 requests, backEndTime: 1.46s, firstPaint: 3.90s, firstVisualChange: 67ms, DOMContentLoaded: 3.71s, Load: 3.71s, speedIndex: 108, visualComplete85: 167ms, lastVisualChange: 167ms, rumSpeedIndex: 3896
[2020-02-26 22:35:03] INFO: HTML stored in /sitespeed.io/sitespeed-result/perf-test_js/2020-02-26-22-34-33
```

**Comment by soulgalore** [comment]
Hi @grantis by default Graphite have a very conservative number of new metrics per minute configured (something like 50) and we are sending a couple of thousand, so if you run multiple times, it will start to add more and more metrics. If you run once you should at least be able to see that some metrics are there. Are you used to use Grafana? Then you can navigate through the namespace. If **sitespeed** exists at your top level, then your metrics are getting there.

**Comment by grantis** [comment]
Hi @soulgalore thought id give an update. I got it posting to graphite, not sure if this is the recommended way so thought id share if someone else has had the trouble iv had. I was unable to get it working with the config options so im manually posting with wget. 

I had trouble finding the start script for the sitespeedio/browsertime or sitespeedio/sitespeed.io docker images so maybe that could be easier for folks. Also noticed sitespeed is creating a folder based on the path so it was important to use the --resultDir so i didn't get a folder name like '.-test-perf-my-test.js' coz thats a bad folder name haha. Also alot of this is new for me so sorry If I'm not doing it correctly ☮️ Appreciate the support 

```
sudo apt-get install jq -y

# Generating Report
/usr/src/app/docker/scripts/start.sh ./test/perf/sitespeed.js --browser "chrome" --resultDir report --skipHar true --iterations 1 --prettyPrint true --video false --visualMetrics false
      
PERF_REPORT=$(cat ./report/browsertime.json | jq '.[0].cdp.performance[0].TaskDuration')

# Send Report to Graphite
TIMESTAMP=$(date +%s)
BASIC_TOKEN=$(echo -n "username:shhhhhsecret" | base64 | tr -d '\n')

wget -qO- \
      --method POST \
      --header "content-type: application/json" \
      --header "Authorization: Basic $BASIC_TOKEN" \
      --body-data "[{\"name\": \"perf.login\", \"interval\": 1, \"value\": $PERF_REPORT , \"time\": $TIMESTAMP }]"  \
      "https://graphite-us-central1.grafana.net/metrics"
```

**Comment by AntoniusGolly** [comment]
Thanks @grantis for posting your solution. It is a bit unfortunate to not get it done by the config. 

I have the same problem. I don't get any more errors but also don't receive anything in Grafana Cloud. 

May I ask how you generate the report? Where did you put the above posted script? I guess it is a script run inside the docker image. But can you elaborate a bit more? It would be extremely helpful.

**Comment by soulgalore** [comment]
Hi, my guess is that first step should be to follow https://grafana.com/docs/grafana-cloud/metrics/graphite/ and install Carbon-Relay-NG and run that. That looks like the easiest way to use Graphite cloud. Else we need to add support for the [HTTP API](https://grafana.com/docs/grafana-cloud/metrics/graphite/http-api/)  it in the Graphite plugin for sitespeed.io.

**Comment by AntoniusGolly** [comment]
@soulgalore Alright. I can do that. Any hint how to automize it? When I use the default command like

```docker run --rm -v "$(pwd):/sitespeed.io" sitespeedio/sitespeed.io:16.2.1 --config=config.json urls.txt```

all the results are in that container. I could use a hint or documentation on where to hook in to send the data.

**Comment by soulgalore** [comment]
If you use the config for graphite (--graphite.host etc) and point that to the Carbon-Relay-NG instance I think that should do the job?

**Comment by AntoniusGolly** [comment]
I think where I'm struggling is how to point that to the Carbon-Relay-NG instance from *within* the sitespeed container. The sidespeed container is a very short-living container as the command above instructs. How could I get started?

**Comment by soulgalore** [comment]
Ok, let me try to explain it better. You setup your Carbon-Relay-NG somewhere (preferable not on the same server that you run sitespeed.io because you want to have as little as possible that can disturb your metrics). Use the host name of the server that runs Carbon-Relay-NG  and set that that host name in ` --graphite.host ` or if you use the json config file, set it there. Then sitesitpeed will send the metrics to the relay and the relay will send it to your Graphite cloud instance.

**Comment by AntoniusGolly** [comment]
That makes total sense. Thanks for the outlined solution.

**Comment by soulgalore** [comment]
@AntoniusGolly did you get it to work? I wanna add a documentation section about setting up Grafana Cloud and wanted to check if you have any gotchas or something I should know.

**Comment by AntoniusGolly** [comment]
I'm at it following your [suggestion](https://github.com/sitespeedio/sitespeed.io/issues/2894#issuecomment-758726821) . How could I provide my findings? Can I create a PR to the docs repo simply?

**Comment by soulgalore** [comment]
@AntoniusGolly yes that would be great! Maybe it could be a section on https://github.com/sitespeedio/sitespeed.io/edit/main/docs/documentation/sitespeed.io/graphite/index.md? Let me know if you need any help.

**Comment by AntoniusGolly** [comment]
Ok, I'll do that. 
It was actually quite a long journey. Easy steps in particular but many of them (installations, configs, firewalls, etc). I got it finally to work. ONCE. I got literally *one* data point into my Grafan Dashboard via Grafana Cloud. And then not any follow-up data point. Obviously, I checked `storage-schemas.conf` of carbon-relay-ng to decrease the interval to 10s, to no avail. I think I need to start from scratch... But I will definitely document that journey. Let me know if you have any insights
